File: backend/app/models/portfolio.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

Realm = Literal["a-share", "etf"]


class PortfolioMember(BaseModel):
    symbol: str
    added_at: str = ""
    note: str = ""


class PortfolioRecord(BaseModel):
    id: str
    name: str
    realm: Realm = "a-share"
    members: list[PortfolioMember] = Field(default_factory=list)
    created_at: str = ""
    updated_at: str = ""


class PortfolioCreate(BaseModel):
    id: str = Field(min_length=1, max_length=48)
    name: str = Field(min_length=1, max_length=80)
    realm: Realm = "a-share"
    members: list[PortfolioMember] = Field(default_factory=list)


class PortfolioUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=80)
    members: list[PortfolioMember] | None = None


class PortfolioMemberSnapshot(BaseModel):
    symbol: str
    name: str | None = None
    note: str | None = None
    price: float | None = None
    pct_change: float | None = None
    chg_5d: float | None = None
    chg_20d: float | None = None


class PortfolioSnapshot(BaseModel):
    portfolio_id: str
    name: str
    realm: Realm = "a-share"
    as_of: str
    member_count: int
    equal_weight_pct_1d: float | None = None
    equal_weight_chg_5d: float | None = None
    equal_weight_chg_20d: float | None = None
    members: list[PortfolioMemberSnapshot] = Field(default_factory=list)
    best_today: PortfolioMemberSnapshot | None = None
    worst_today: PortfolioMemberSnapshot | None = None


class TrackRecord(BaseModel):
    id: str
    date: str
    recorded_at: str
    equal_weight_pct_1d: float | None = None
    equal_weight_chg_5d: float | None = None
    member_count: int = 0
    best_symbol: str | None = None
    best_name: str | None = None
    best_pct: float | None = None
    worst_symbol: str | None = None
    worst_name: str | None = None
    worst_pct: float | None = None
    note: str | None = None
    path_id: str | None = None


class TrackRecordCreate(BaseModel):
    note: str | None = None
    path_id: str | None = None

File: frontend/src/types/portfolio.ts
export type PortfolioRealm = "a-share" | "etf";

export interface PortfolioMember {
  symbol: string;
  added_at?: string;
  note?: string;
}

export interface PortfolioRecord {
  id: string;
  name: string;
  realm: PortfolioRealm;
  members: PortfolioMember[];
  created_at?: string;
  updated_at?: string;
}

export interface PortfolioMemberSnapshot {
  symbol: string;
  name?: string | null;
  note?: string | null;
  price?: number | null;
  pct_change?: number | null;
  chg_5d?: number | null;
  chg_20d?: number | null;
}

export interface PortfolioSnapshot {
  portfolio_id: string;
  name: string;
  realm: PortfolioRealm;
  as_of: string;
  member_count: number;
  equal_weight_pct_1d?: number | null;
  equal_weight_chg_5d?: number | null;
  equal_weight_chg_20d?: number | null;
  members: PortfolioMemberSnapshot[];
  best_today?: PortfolioMemberSnapshot | null;
  worst_today?: PortfolioMemberSnapshot | null;
}

export interface TrackRecord {
  id: string;
  date: string;
  recorded_at: string;
  equal_weight_pct_1d?: number | null;
  equal_weight_chg_5d?: number | null;
  member_count: number;
  best_symbol?: string | null;
  best_name?: string | null;
  best_pct?: number | null;
  worst_symbol?: string | null;
  worst_name?: string | null;
  worst_pct?: number | null;
  note?: string | null;
  path_id?: string | null;
}

export interface PortfolioUpdateBody {
  name?: string;
  members?: PortfolioMember[];
}
